File: tests/algorithms/x/Java/dynamic_programming/largest_divisible_subset.java
public class Main {

    static int[] sort_list(int[] nums) {
        int[] arr = ((int[])(nums));
        int i = 1;
        while (i < arr.length) {
            int key = arr[i];
            int j = i - 1;
            while (j >= 0 && arr[j] > key) {
arr[j + 1] = arr[j];
                j = j - 1;
            }
arr[j + 1] = key;
            i = i + 1;
        }
        return arr;
    }

    static int[] largest_divisible_subset(int[] items) {
        if (items.length == 0) {
            return new int[]{};
        }
        int[] nums = ((int[])(sort_list(((int[])(items)))));
        int n = nums.length;
        int[] memo = ((int[])(new int[]{}));
        int[] prev = ((int[])(new int[]{}));
        int i_1 = 0;
        while (i_1 < n) {
            memo = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(memo), java.util.stream.IntStream.of(1)).toArray()));
            prev = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(prev), java.util.stream.IntStream.of(i_1)).toArray()));
            i_1 = i_1 + 1;
        }
        i_1 = 0;
        while (i_1 < n) {
            int j_1 = 0;
            while (j_1 < i_1) {
                if ((nums[j_1] == 0 || Math.floorMod(nums[i_1], nums[j_1]) == 0) && memo[j_1] + 1 > memo[i_1]) {
memo[i_1] = memo[j_1] + 1;
prev[i_1] = j_1;
                }
                j_1 = j_1 + 1;
            }
            i_1 = i_1 + 1;
        }
        int ans = 0 - 1;
        int last_index = 0 - 1;
        i_1 = 0;
        while (i_1 < n) {
            if (memo[i_1] > ans) {
                ans = memo[i_1];
                last_index = i_1;
            }
            i_1 = i_1 + 1;
        }
        if (last_index == 0 - 1) {
            return new int[]{};
        }
        int[] result = ((int[])(new int[]{nums[last_index]}));
        while (prev[last_index] != last_index) {
            last_index = prev[last_index];
            result = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(result), java.util.stream.IntStream.of(nums[last_index])).toArray()));
        }
        return result;
    }

    static void main() {
        int[] items = ((int[])(new int[]{1, 16, 7, 8, 4}));
        int[] subset = ((int[])(largest_divisible_subset(((int[])(items)))));
        System.out.println("The longest divisible subset of " + _p(items) + " is " + _p(subset) + ".");
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            main();
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/x/Pascal/dynamic_programming/largest_divisible_subset.pas
{$mode objfpc}
program Main;
uses SysUtils;
type IntArray = array of integer;
type IntArrayArray = array of IntArray;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
function list_int_to_str(xs: array of integer): string;
var i: integer;
begin
  Result := '[';
  for i := 0 to High(xs) do begin
    Result := Result + IntToStr(xs[i]);
    if i < High(xs) then Result := Result + ' ';
  end;
  Result := Result + ']';
end;
function list_list_int_to_str(xs: array of IntArray): string;
var i: integer;
begin
  Result := '[';
  for i := 0 to High(xs) do begin
    Result := Result + list_int_to_str(xs[i]);
    if i < High(xs) then Result := Result + ' ';
  end;
  Result := Result + ']';
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  items: IntArray;
  nums: IntArray;
function sort_list(nums: IntArray): IntArray; forward;
function largest_divisible_subset(items: IntArray): IntArray; forward;
procedure main(); forward;
function sort_list(nums: IntArray): IntArray;
var
  sort_list_arr: array of integer;
  sort_list_i: integer;
  sort_list_key: integer;
  sort_list_j: integer;
begin
  sort_list_arr := nums;
  sort_list_i := 1;
  while sort_list_i < Length(sort_list_arr) do begin
  sort_list_key := sort_list_arr[sort_list_i];
  sort_list_j := sort_list_i - 1;
  while (sort_list_j >= 0) and (sort_list_arr[sort_list_j] > sort_list_key) do begin
  sort_list_arr[sort_list_j + 1] := sort_list_arr[sort_list_j];
  sort_list_j := sort_list_j - 1;
end;
  sort_list_arr[sort_list_j + 1] := sort_list_key;
  sort_list_i := sort_list_i + 1;
end;
  exit(sort_list_arr);
end;
function largest_divisible_subset(items: IntArray): IntArray;
var
  largest_divisible_subset_nums: IntArray;
  largest_divisible_subset_n: integer;
  largest_divisible_subset_memo: array of integer;
  largest_divisible_subset_prev: array of integer;
  largest_divisible_subset_i: integer;
  largest_divisible_subset_j: integer;
  largest_divisible_subset_ans: integer;
  largest_divisible_subset_last_index: integer;
  largest_divisible_subset_result_: array of integer;
begin
  if Length(items) = 0 then begin
  exit([]);
end;
  largest_divisible_subset_nums := sort_list(items);
  largest_divisible_subset_n := Length(largest_divisible_subset_nums);
  largest_divisible_subset_memo := [];
  largest_divisible_subset_prev := [];
  largest_divisible_subset_i := 0;
  while largest_divisible_subset_i < largest_divisible_subset_n do begin
  largest_divisible_subset_memo := concat(largest_divisible_subset_memo, IntArray([1]));
  largest_divisible_subset_prev := concat(largest_divisible_subset_prev, IntArray([largest_divisible_subset_i]));
  largest_divisible_subset_i := largest_divisible_subset_i + 1;
end;
  largest_divisible_subset_i := 0;
  while largest_divisible_subset_i < largest_divisible_subset_n do begin
  largest_divisible_subset_j := 0;
  while largest_divisible_subset_j < largest_divisible_subset_i do begin
  if ((largest_divisible_subset_nums[largest_divisible_subset_j] = 0) or ((largest_divisible_subset_nums[largest_divisible_subset_i] mod largest_divisible_subset_nums[largest_divisible_subset_j]) = 0)) and ((largest_divisible_subset_memo[largest_divisible_subset_j] + 1) > largest_divisible_subset_memo[largest_divisible_subset_i]) then begin
  largest_divisible_subset_memo[largest_divisible_subset_i] := largest_divisible_subset_memo[largest_divisible_subset_j] + 1;
  largest_divisible_subset_prev[largest_divisible_subset_i] := largest_divisible_subset_j;
end;
  largest_divisible_subset_j := largest_divisible_subset_j + 1;
end;
  largest_divisible_subset_i := largest_divisible_subset_i + 1;
end;
  largest_divisible_subset_ans := 0 - 1;
  largest_divisible_subset_last_index := 0 - 1;
  largest_divisible_subset_i := 0;
  while largest_divisible_subset_i < largest_divisible_subset_n do begin
  if largest_divisible_subset_memo[largest_divisible_subset_i] > largest_divisible_subset_ans then begin
  largest_divisible_subset_ans := largest_divisible_subset_memo[largest_divisible_subset_i];
  largest_divisible_subset_last_index := largest_divisible_subset_i;
end;
  largest_divisible_subset_i := largest_divisible_subset_i + 1;
end;
  if largest_divisible_subset_last_index = (0 - 1) then begin
  exit([]);
end;
  largest_divisible_subset_result_ := [largest_divisible_subset_nums[largest_divisible_subset_last_index]];
  while largest_divisible_subset_prev[largest_divisible_subset_last_index] <> largest_divisible_subset_last_index do begin
  largest_divisible_subset_last_index := largest_divisible_subset_prev[largest_divisible_subset_last_index];
  largest_divisible_subset_result_ := concat(largest_divisible_subset_result_, IntArray([largest_divisible_subset_nums[largest_divisible_subset_last_index]]));
end;
  exit(largest_divisible_subset_result_);
end;
procedure main();
var
  main_items: array of integer;
  main_subset: IntArray;
begin
  main_items := [1, 16, 7, 8, 4];
  main_subset := largest_divisible_subset(main_items);
  writeln(((('The longest divisible subset of ' + list_int_to_str(main_items)) + ' is ') + list_int_to_str(main_subset)) + '.');
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  main();
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.
